metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala

```diff
@@ -215,6 +216,101 @@ final class FormattingProvider(
   }
 
+  private def highestDialect(): Option[ScalafmtDialect] =
+    buildTargets.all.toList
+      .map(_.fmtDialect)
+      .sorted
+      .lastOption
+
+  private def initialConfig(): String = {
+    val runnerDialect = highestDialect.filter(_ != ScalafmtDialect.Scala213)
+
+    val fmtVersion =
+      if (runnerDialect.isDefined) BuildInfo.scalafmt3Version
+      else BuildInfo.scalafmtVersion
+
+    val versionLine = s"""version = "$fmtVersion""""
+    val runnerLine = runnerDialect.map(dialect => s"runner.dialect = $dialect")
+    List(
+      Some(versionLine),
+      runnerLine,
+      Some("")
+    ).flatten.mkString("\n")
+  }
+
+  private def scalftmVersionForDialect(d: ScalafmtDialect): String =
+    d match {
+      case Scala213 => BuildInfo.scalafmtVersion
+      case Scala213Source3 | Scala3 => BuildInfo.scalafmt3Version
+    }
+
+  private def upgradeDialect(
+      configValues: ScalafmtConfig.ConfigValues
+  ): Option[ScalafmtDialect] = {
+    val initial = configValues.runnerDialect.getOrElse(ScalafmtDialect.Scala213)
+    val targetDialects = buildTargets.all.toList.map(_.fmtDialect)
+
+    if (targetDialects.nonEmpty) {
+      val highest = targetDialects.max
+      if (highest != initial) {
+        // if the highest dialect was mentioned in overrides do not suggest upgrading
+        val usedInOverrides =
+          configValues.fileOverrides.values.exists(_ == highest)
+        if (usedInOverrides) None else Some(highest)
+      } else None
+    } else None
+  }
+
+  private def checkIfDialectUpgradeRequired(
+      configValues: ScalafmtConfig.ConfigValues
+  ): Future[Unit] = {
+    upgradeDialect(configValues) match {
+      case Some(required)
+          if !tables.dismissedNotifications.UpdateScalaftmConf.isDismissed =>
+        val params = UpdateScalafmtConf.params(required)
+        client.showMessageRequest(params).asScala.map { item =>
+          if (item == UpdateScalafmtConf.letUpdate) {
+            val requiredVersion = scalftmVersionForDialect(required)
+            val updVersion =
+              if (!configValues.version.exists(_ == requiredVersion))
+                Some(requiredVersion)
+              else
+                None
+
+            val text = scalafmtConf.toInputFromBuffers(buffers).text
+            val updatedText =
+              ScalafmtConfig.update(text, updVersion, Some(required))
+
+            Files.write(
+              scalafmtConf.toNIO,
+              updatedText.getBytes(StandardCharsets.UTF_8)
+            )
+          } else if (item == Messages.notNow) {
+            tables.dismissedNotifications.UpdateScalaftmConf
+              .dismiss(24, TimeUnit.HOURS)
+          } else if (item == Messages.dontShowAgain) {
+            tables.dismissedNotifications.UpdateScalaftmConf.dismissForever()
+          } else ()
+        }
+      case _ => Future.unit
+    }
+  }
+
+  def validateWorkspace(): Future[Unit] = {
+    if (scalafmtConf.exists) {
```

Comment:
So I know when we do an actual format we do a check to ensure that if the user has manually set the `scalafmtConfigPath` it actually exists. here we also call `scalafmtConf`, but don't do that check. Do you think we should also do a check to ensure something wonky doesn't happen here, or even move that check into `scalafmtConf` to ensure that not only have they set the configPath, but it's also valid and there?

Reply:
`scalafmtConf` is this function that checks `userConfig`
https://github.com/scalameta/metals/blob/270804287a8abe18bdd072ec7cb0608fc77b1e2c/metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala#L362-L365

Reply:
Sure, but it doesn't validate anything right? So if `userConfig().scalafmtConfigPath` returns `/my/wrong/path.conf`, that is still returned here and give to you even if it doesn't exists.

Reply:
oh. now I got what you mean. Will fix that

Reply:
Did two changes around that:
- in case if the conf file doesn't exist or has syntax error now message shows the correct file path in the text
- in case if upgrade is required Metals will suggest only to fix if it's the default one - (`.scalafmt.conf`).
  For a file that is specified in `userConfig` it will only warn about that as it might be shared across several projects.